### PR TITLE
C++20 move_iterator changes and move_sentinel

### PR DESCRIPTION
**Pull request by CaseyCarter**
Changes primarily come from P0896R4 "`<ranges>`", but the `operator<=>` from P1614R2 "Adding Spaceship `<=>` To The Library" is here as well.

I've also speculatively implemented the proposed resolutions of LWG-3293, LWG-3391, and LWG-3435.

I had to [patch a pair of libc++ tests](https://reviews.llvm.org/D79343) that fail with these changes; I'll update the LLVM reference herein once those changes merge upstream. (These are the failing tests.)

Drive-by:
* s/explicit constexpr/constexpr explicit/g (I somehow managed to sneak two occurrences into `subrange`)

### CODE REVIEW

File: stl/inc/iterator [hunk 1]
```diff
@@ -186,6 +186,44 @@ _NODISCARD insert_iterator<_Container> inserter(_Container& _Cont, typename _Con
...
+
+    constexpr explicit move_sentinel(_Se _Val) noexcept(is_nothrow_move_constructible_v<_Se>) // strengthened
+        : _Last(_STD move(_Val)) {}
+
+    // clang-format off
+    template <class _Se2>
+        requires convertible_to<const _Se2&, _Se>
+    constexpr move_sentinel(const move_sentinel<_Se2>& _Val)
+        noexcept(is_nothrow_constructible_v<_Se, const _Se2&>) // strengthened
+        : _Last{_Val._Get_last()} {}
```
Comment (StephanTLavavej) [inline stl/inc/iterator:204]: I observe that you're using braces here, but parens on line 197. Should they be consistent?

File: stl/inc/iterator [hunk 2]
```diff
@@ -186,6 +186,44 @@ _NODISCARD insert_iterator<_Container> inserter(_Container& _Cont, typename _Con
...
+    template <class _Se2>
+        requires assignable_from<_Se&, const _Se2&>
+    constexpr move_sentinel& operator=(const move_sentinel<_Se2>& _Val)
+        noexcept(is_nothrow_assignable_v<_Se&, const _Se2&>) /* strengthened */ {
+        _Last = _Val._Get_last();
+        return *this;
+    }
+    // clang-format on
+
+    constexpr _Se base() const noexcept(is_nothrow_copy_constructible_v<_Se>) /* strengthened */ {
```
Comment (StephanTLavavej) [inline stl/inc/iterator:215]: Should this be `_NODISCARD` as a pure observer?

File: stl/inc/iterator [hunk 3]
```diff
@@ -186,6 +186,44 @@ _NODISCARD insert_iterator<_Container> inserter(_Container& _Cont, typename _Con
...
+        return _Last;
+    }
+
+    constexpr const _Se& _Get_last() const noexcept {
```
Comment (StephanTLavavej) [inline stl/inc/iterator:219]: Ditto `_NODISCARD` (which we're now trying to use on internal machinery)?

File: stl/inc/xutility [hunk 4]
```diff
@@ -3695,39 +3695,82 @@ struct _Iterator_base12;
 struct _Default_sentinel {}; // empty struct to serve as the end of a range
 
 // CLASS TEMPLATE move_iterator
+#ifdef __cpp_lib_concepts
+template <semiregular>
```
Comment (StephanTLavavej) [inline stl/inc/xutility:3699]: Should we include a parameter name in this forward declaration?

File: stl/inc/xutility [hunk 5]
```diff
@@ -3765,34 +3816,69 @@ public:
...
+#ifdef __cpp_lib_concepts
+        return _RANGES iter_move(current + _Off);
+#else // ^^^ __cpp_lib_concepts / !__cpp_lib_concepts vvv
         return _STD move(current[_Off]);
+#endif // __cpp_lib_concepts
+    }
+
+#ifdef __cpp_lib_concepts
+    template <sentinel_for<_Iter> _Sent>
+    friend constexpr bool operator==(const move_iterator& _Left, const move_sentinel<_Sent>& _Right) {
```
Comment (StephanTLavavej) [inline stl/inc/xutility:3847]: Looks like these operators should be `_NODISCARD`.

File: stl/inc/xutility [hunk 6]
```diff
@@ -3802,63 +3888,111 @@ public:
...
-    return !(_Right < _Left);
+_NODISCARD _CONSTEXPR17 bool operator>=(const move_iterator<_Iter1>& _Left, const move_iterator<_Iter2>& _Right)
+#ifdef __cpp_lib_concepts
+    // clang-format off
+    requires requires { _Left < _Right; }
+// clang-format on
+#endif // __cpp_lib_concepts
+{ return !(_Left < _Right); }
+
+#if defined(__cpp_impl_three_way_comparison) && defined(__cpp_lib_concepts)
```
Comment (miscco) [inline stl/inc/xutility:3970]: Thinking out loud. Should this have an else clause that encloses everything above from `operator==` onwards.

The individual definitions should not be necessary with spaceship

File: stl/inc/xutility [hunk 7]
```diff
@@ -3736,10 +3779,18 @@ public:
         return *this;
     }
 
-    _CONSTEXPR17 move_iterator operator++(int) {
-        move_iterator _Tmp = *this;
-        ++current;
-        return _Tmp;
+    _CONSTEXPR17 auto operator++(int) {
```
Comment (miscco) [inline stl/inc/xutility:3782]: All this duplication. We finally need operator spacestation `++--+-<=>-+--++`

File: stl/inc/iterator (hunk 2 above)
Comment (CaseyCarter) [reply stl/inc/iterator:215]: Yes!

File: stl/inc/xutility (hunk 4 above)
Comment (CaseyCarter) [reply stl/inc/xutility:3699]: Unreferenced names are at best documentation; I see no need to document this forward declaration of a standardized library entity. Folks who care can "Peek Definition" or look at cppreference.

Comment (StephanTLavavej) [reply stl/inc/xutility:3699]: Ok.

File: stl/inc/xutility (hunk 6 above)
Comment (CaseyCarter) [reply stl/inc/xutility:3970]: `move_iterator`'s `<=>` is only provided when the adapted iterator provides `<=>`, i.e., it requires `three_way_comparable_with`. We still need these individual operators when the adaptee doesn't provide `<=>`, or when the adaptee has relational operators that are inconsistent with `operator<=>`.

**Comment by CaseyCarter** [comment]
Thanks for this... :sunglasses: ...moving contribution.

**Pull request merged**